Index: libogc/ipc.c
===================================================================
--- libogc/ipc.c	(revision 3368)
+++ libogc/ipc.c	(working copy)
@@ -856,7 +856,7 @@
 	req->cb = NULL;
 	req->relnch = 0;
 
-	DCFlushRange((void*)filepath,strnlen(filepath,IPC_MAXPATH_LEN));
+	DCFlushRange((void*)filepath,strnlen(filepath,IPC_MAXPATH_LEN) + 1);
 
 	req->open.filepath	= (char*)MEM_VIRTUAL_TO_PHYSICAL(filepath);
 	req->open.mode		= mode;

