Index: libogc/network_wii.c
===================================================================
--- libogc/network_wii.c	(revision 3498)
+++ libogc/network_wii.c	(working copy)
@@ -53,7 +53,7 @@
 #include "network.h"
 #include "ogcsys.h"
 
-#define NET_HEAP_SIZE				8192
+#define NET_HEAP_SIZE				34816
 
 #define IOS_O_NONBLOCK				0x04			//(O_NONBLOCK >> 16) - it's in octal representation, so this shift leads to 0 and hence nonblocking sockets didn't work. changed it to the right value.
 
Index: libogc/wiisd.c
===================================================================
--- libogc/wiisd.c	(revision 3498)
+++ libogc/wiisd.c	(working copy)
@@ -112,6 +112,7 @@
 
 #define	SDIO_STATUS_CARD_INSERTED	0x1
 #define	SDIO_STATUS_CARD_INITIALIZED	0x10000
+#define	SDIO_STATUS_CARD_SDHC	    0x100000
  
 #define READ_BL_LEN					((u8)(__sd0_csd[5]&0x0f))
 #define WRITE_BL_LEN				((u8)(((__sd0_csd[12]&0x03)<<2)|((__sd0_csd[13]>>6)&0x03)))
@@ -455,6 +456,8 @@
 		ret = __sd0_getrca();
 		if(ret < 0) goto fail;
 	}
+	else if (status & SDIO_STATUS_CARD_SDHC)
+		__sd0_sdhc = 1;
 	else
 		__sd0_sdhc = 0;
  
