Index: source/disc.c
===================================================================
--- source/disc.c	(revision 3145)
+++ source/disc.c	(working copy)
@@ -41,16 +41,25 @@
 /* ====================== Wii ====================== */
 #if   defined (__wii__)
 #include <sdcard/wiisd_io.h>
+#include <sdcard/gcsd.h>
 #include <ogc/usbstorage.h>
 
 static const DISC_INTERFACE* get_io_wiisd (void) {
 	return &__io_wiisd;
 }
+static const DISC_INTERFACE* get_io_gcsda (void) {
+	return &__io_gcsda;
+}
+static const DISC_INTERFACE* get_io_gcsdb (void) {
+	return &__io_gcsdb;
+}
 static const DISC_INTERFACE* get_io_usbstorage (void) {
 	return &__io_usbstorage;
 }
 const INTERFACE_ID _FAT_disc_interfaces[] = {
 	{"sd", get_io_wiisd},
+	{"gcsda", get_io_gcsda},
+	{"gcsdb", get_io_gcsdb},
 	{"usb", get_io_usbstorage},
 	{NULL, NULL}
 };	
